server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -366,6 +401,242 @@ public ClusterState getLastAcceptedState() {
+                            : "onPossibleCompletion: term or version mismatch when publishing [" + this
+                            + "]: current version is now [" + coordinationState.get().getLastAcceptedVersion()
+                            + "] in term [" + coordinationState.get().getLastAcceptedTerm() + "]";
+
+                        // TODO: send to applier
+                        ackListener.onNodeAck(getLocalNode(), null);
+                        completionListener.onResponse(null);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+                        assert committed == false;
+                        if (publishRequest.getAcceptedState().term() == coordinationState.get().getCurrentTerm() &&
+                            publishRequest.getAcceptedState().version() == coordinationState.get().getLastPublishedVersion()) {
+                            becomeCandidate("Publication.onCompletion(false)");
+                        }
+                        FailedToCommitClusterStateException exception = new FailedToCommitClusterStateException("publication failed", e);
+                        ackListener.onNodeAck(getLocalNode(), exception); // other nodes have acked, but not the master.
+                        completionListener.onFailure(exception);
+                    }
+                }, transportService.getThreadPool().generic());
+            }
+
+            @Override
+            protected boolean isPublishQuorum(CoordinationState.VoteCollection votes) {
+                assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+                return coordinationState.get().isPublishQuorum(votes);
+            }
+
+            @Override
+            protected Optional<ApplyCommitRequest> handlePublishResponse(DiscoveryNode sourceNode, PublishResponse publishResponse) {
+                assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+                assert getCurrentTerm() >= publishResponse.getTerm();
+                return coordinationState.get().handlePublishResponse(sourceNode, publishResponse);
+            }
+
+            @Override
+            protected void onJoin(Join join) {
+                assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+                if (join.getTerm() == getCurrentTerm()) {
+                    handleJoin(join);
+                }
+                // TODO: what to do on missing join?
+            }
+
+            @Override
+            protected void sendPublishRequest(DiscoveryNode destination, PublishRequest publishRequest,
+                                              ActionListener<PublishWithJoinResponse> responseActionListener) {
+                transportService.sendRequest(destination, PUBLISH_STATE_ACTION_NAME, publishRequest,
```

Comment:
Is it possible to move this out of `Coordinator`? Maybe all the way into `Publication`, and update the tests to use a `Transport` always rather than overriding these methods?

Reply:
There is more than just the transport call, namely wrapping the response handler with the mutex. I've factored this into a PublicationTransportHandler. This class will later on also do the diff-based publishing.